src/main/scala/com/mwz/sonar/scala/scapegoat/ScapegoatReportParser.scala

```diff
@@ -0,0 +1,53 @@
+/*
+ * Sonar Scala Plugin
+ * Copyright (C) 2018 All contributors
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package com.mwz.sonar.scala.scapegoat
+
+import java.nio.file.Path
+import scala.xml.XML
+
+/** Used to mock the scoverage report parser in tests */
+private[scapegoat] trait ScapegoatReportParserAPI {
+
+  /** Parses the scapegoat xml report and returns all scapegoat warnings by filename */
+  def parse(scapegoatReportPath: Path): Map[String, Seq[ScapegoatWarning]]
+}
+
+/** Scapegoat XML reports parser */
+private[scapegoat] trait ScapegoatReportParser extends ScapegoatReportParserAPI {
```

Comment:
I'm generally not a fan of the 🍰 pattern and I prefer to use constructor arguments instead. It's just a suggestion to consider, that if you were to make the `ScapegoatReportParser` a class or even if you instantiated the trait you could pass it down as an argument to the `ScapegoatSensor` class and you could avoid having to deal with all of those abstract classes and cake pattern alltogether, which makes it much easier to write unit tests.

You can read about it here https://github.com/alexandru/scala-best-practices/blob/master/sections/3-architecture.md#31-should-not-use-the-cake-pattern.

Reply:
Oh yeah yeah, I really know that the _infamous_ cake pattern can become a mess.

However, there is a reason why I use it.
Remember we don't instantiate the sensor by ourselves - it is SonarQube who does it.
And I tried to make it injectable dependency, but when I tested it, sonar failed to instantiate the sensor.

That problem comes from the **ScoverageSensor**. _(I think we already discussed it, but not sure)_.

The good point is, it's not a complicated cake pattern. So not so bad.

But maybe I'm missing something from the SonarQube documentation.
I'll open an Issue to look into this with more detail later.

Reply:
I see, but when you look at the old Scalastyle sensor, all of the dependencies are injected via constructor arguments, I think the issue in your case was having the `ScapegoatReportParser` scoped to the scapegoat package so it's invisible from outside and Sonar fails to inject it.

So I was thinking about something along those lines:

```scala
trait ScapegoatReportParserAPI { ...
```
```scala
final class ScapegoatReportParser extends ScapegoatReportParserAPI { ...
```
and then the sensor can be simply defined like that:
```scala
final class ScapegoatSensor(scapegoatReportParser: ScapegoatReportParser) extends Sensor { ...
```

I think tht should work.

Reply:
Oh, You may be correct.
If it works for you I'll do the same with Scapegoa. :+1: